R/R/prophet.R

```diff
@@ -1274,37 +1268,203 @@ fit.prophet <- function(m, df, ...) {
+
+#' Output a list of all data required for stan fitting or prediction
+#'
+#' @param m The fitted prophet object
+#' @param df Data frame with historical or future data for fitting or prediction resp.
+#' @param fit.seasonal.features matrix of seasonal, holiday & exogenous regressors for fitting
+#' @param predict.seasonal.features  matrix of seasonal, holiday & exogenous regressors for predictions
+#' @param component.cols indicators for additive or multiplicative components in seasonal.features
+#' @param prior.scales prior scale hyperparameters
+#' @param type 'fit' or 'predict'
+#' 
+#' @return List with all data required for stan fitting or prediction
+#'
+#' @keywords internal
+data_for_stan <- function(
+  m, df, 
+  fit.seasonal.features, 
+  predict.seasonal.features = matrix(0, 0, ncol(fit.seasonal.features)),
+  component.cols, 
+  prior.scales, 
+  type) {
+  
+  if (type == "fit") {
+    
+    dat <- list(
+      T = nrow(df),
+      K = ncol(fit.seasonal.features),
+      S = length(m$changepoints.t),
+      y = df$y_scaled,
+      t = df$t,
+      cap = switch(
+        m$growth, 
+        linear = rep(0, nrow(df)),                     # Unused inside Stan
+        df$cap_scaled                                  # Add capacities to the Stan data
+      ), 
+      t_change = array(m$changepoints.t),
+      X = as.matrix(fit.seasonal.features),
+      sigmas = array(prior.scales),
+      tau = m$changepoint.prior.scale,
+      trend_indicator = as.numeric(m$growth == 'logistic'),
+      s_a = array(component.cols$additive_terms),
+      s_m = array(component.cols$multiplicative_terms),
+      
+      T_pred = 0,                                      #dummy values
+      t_pred = rep(0, 0),                              #dummy values
+      cap_pred = rep(0, 0),                            #dummy values
+      X_pred = as.matrix(predict.seasonal.features),   #dummy values
+      n_samp = 1,                                      #dummy values
+      S_pred = 1                                       #dummy values
+    )
+    
+  } else {
+    
+    dat <- list(
+      T = nrow(m$history),
+      K = ncol(fit.seasonal.features),
+      S = length(m$changepoints.t),
+      y = m$history$y_scaled,                         
+      t = m$history$t,                                
+      cap = switch(
+        m$growth, 
+        linear = rep(0, nrow(m$history)),              # Unused inside Stan
+        m$history$cap_scaled                           # Add capacities to the Stan data
+      ),
+      t_change = array(m$changepoints.t),
+      X = as.matrix(fit.seasonal.features),            
+      sigmas = array(prior.scales),
+      tau = m$changepoint.prior.scale,
+      trend_indicator = as.numeric(m$growth == 'logistic'),
+      s_a = array(component.cols$additive_terms),
+      s_m = array(component.cols$multiplicative_terms),
+      
+      T_pred = nrow(df),
+      t_pred = df$t,
+      cap_pred = 
+        if(!'cap' %in% colnames(df)) { 
```

Comment:
This won't be necessary. If growth='linear', then we would have already raised an error if df did not have a `cap` column.